unity/CAVE/Assets/W3D_Translator/CLI.cs

```diff
@@ -241,25 +251,76 @@ private static void BuildWalls(Story xml, Transform storyT)
                 button.targetGraphic = contentGO.GetComponent<Graphic>(); // Text, Image, etc.
                 button.colors = link.SetColors(button.colors, xml.ColorString);
-                if (!link.RemainEnabled)
-                { // TODO: Disable button after click
-                }
-
-                // TODO (83): Add button actions
-                foreach (LinkActions action in link.Actions)
-                {
-                }
             }
             else
             {
                 contentGO.SetActive(xml.Visible);
                 xml.Placement.SetTransform(contentGO.transform, xml.GetScale(), story.transform);
             }
-            gameObjects.Add(contentGO.name, contentGO);
+            gameObjects.Add(contentGO.name, (contentGO, xml));
         }
         return gameObjects;
     }
 
+    private static void SetLinkActions(ObjDictionary gameObjects, GameObject story)
+    {
+        ActionMethods methods = story.GetComponent<ActionMethods>();
+
+        foreach (KeyValuePair<string, (GameObject, W3D.Object)> pair in
+            gameObjects.Where(pair => pair.Value.Item2.LinkRoot is not null)
+        )
+        {
+            (GameObject go, W3D.Object obj) = pair.Value;
+            GameObject buttonGO = go.transform.parent.gameObject;
+            Button button = buttonGO.GetComponent<Button>();
+            Link link = obj.LinkRoot.Link;
+
+            // Add the <Action>s to onClick
+            Button.ButtonClickedEvent onClick = button.onClick;
+            foreach (LinkActions xmlAction in link.Actions)
+            {
+                if (
+                    xmlAction.Clicks is not null &&
+                    xmlAction.Clicks.Type == Clicks.ActivationTypes.Number
+                )
+                {
+                    // Button is only activated after a certain number of clicks
+                }
+
+                // TODO (83): Add button actions
+                switch (xmlAction.Action)
+                {
+                    case ObjectChange objectRef:
+                        // TODO 86
+                        break;
+                    case GroupRef groupRef:
+                        // TODO 87
+                        break;
+                    case TimerChange timelineRef:
+                        // TODO 88
+                        break;
+                    case W3D.Event eventTriggerRef:
+                        // TODO 89
+                        break;
+                    case MoveCave moveCave:
+                        // TODO 90
+                        break;
+                    case Reference soundRef:
+                        // TODO 91
+                        break;
+                    default:
+                        if (xmlAction.Type == Actions.ActionTypes.Restart)
+                        {
+                            // TODO 92
+                        }
+                        else { throw new Exception("TODO"); }
+                        break;
+                }
```

Comment:
Should the actions be within the if statement above? We only want an action to happen if the right number of clicks happens right?

Reply:
I'm 99% sure no, although I'm not entirely sure how the number of clicks is going to play out.

Here we're just adding callback functions to the buttons onClick, similar to web dev. The difference being that in Unity `onClick` is a list - you can add as many callback functions as you'd like. The actual code looks like 

```c#
            UnityEventTools.AddObjectPersistentListener(
                onClick, // list of callback functions
                new UnityAction<Button>(methods.DisableButton), // Callback function to add
                button // Parameter of methods.DisableButton
            );
```

The question remains as to what will go in the `if` block - how we'll tell Unity to fire the onClick. It's looking a little like I'd add a counter function to the onClick itself and manage the callback functions in a different list? Or I'd capture `onClick.Invoke()` and manage that based on the counter? Not sure how they'll play out but the list of callback functions and how/when they're execute definitely seem to be different tasks

Reply:
https://gamedev.stackexchange.com/questions/163317/trigger-the-buttons-onclick-after-checking-that-the-cursor-has-been-on-it-for-s

https://stackoverflow.com/questions/56323774/unity-how-can-i-make-a-function-fire-only-when-player-clicked-on-it-3-times